### PR TITLE
test/e2e/node + test/e2e/scheduling: improve checks

### PR DESCRIPTION
**Pull request by Ahmed-Aghadi**
#### What type of PR is this?
Enhancement

#### What this PR does / why we need it:
Better output for developers when tests fail.
#### Which issue(s) this PR fixes:

Fixes # Part of https://github.com/kubernetes/kubernetes/issues/105678

#### Does this PR introduce a user-facing change?
None

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by wgahnagl** [comment]
/close

Thank you for your PR. It seems to only contain trivial edits. Please read our [trivial edits policy](https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#trivial-edits). We encourage you to take a look at confirmed issues and bugs or issues marked `help-wanted`.

**Comment by Ahmed-Aghadi** [comment]
> ed issues

Thanks for the reply. But as mentioned in the  PR ( [https://github.com/kubernetes/kubernetes/issues/105678](url) )

> It makes sense to split this work into individual PRs, one per OWNERS file under test/e2e.

I did the part for test/e2e/node and there was very less part in this file. But as it was mentioned in the PR to split it into individual PRs, I had included this one separately. So should I include this change with changes to be done in other section of test/e2e/ ?

**Comment by pohly** [comment]
/reopen

@wgahnagl: while this particular PR is small, it does fix a real issue (https://github.com/kubernetes/kubernetes/issues/105678) and it does so for all files owned by SIG Node under test/e2e. It just happened that there was only one instance of the problem.

File: test/e2e/node/runtimeclass.go
Comment (pohly) [inline test/e2e/node/runtimeclass.go:63]: Include more information here. What is the actual value?

```suggestion
			framework.Failf("expected 'forbidden' as error, got instead: %v", err)
```

File: test/e2e/node/runtimeclass.go
Comment (pohly) [inline test/e2e/node/runtimeclass.go:61]: You can remove this line. `apierrors.IsForbidden` accepts nil and will return false, so the next `Failf` covers both cases.

**Comment by Ahmed-Aghadi** [comment]
@pohly  Thanks a lot for your time. I had updated changes accordingly. Please let me know if further changes are needed.

**Comment by pohly** [comment]
Looks good to me. But before we test it, can you squash into a single commit?

**Comment by pohly** [comment]
@Ahmed-Aghadi : you can also put all changes for test/e2e/scheduling into this PR. Then it will have a good size.

**Comment by Ahmed-Aghadi** [comment]
> @Ahmed-Aghadi : you can also put all changes for test/e2e/scheduling into this PR. Then it will have a good size.

@pohly Sure! I had done for all them except one as I'm unable to find the output for https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scheduling/predicates.go#L1077

**Comment by pohly** [comment]
You can change that into:
```
			_, scheduledCondition := podutil.GetPodCondition(&pod.Status, v1.PodScheduled)
			if scheduledCondition == nil {
			        framework.Failf("Did not find 'scheduled' condition for pod %+v", pod)
			}
			framework.ExpectEqual(scheduledCondition.Status, v1.ConditionTrue)
			scheduledPods = append(scheduledPods, pod)
```

**Comment by Ahmed-Aghadi** [comment]
@pohly I had updated all the changes for test/e2e/scheduling, can you review them?

File: test/e2e/scheduling/events.go
Comment (pohly) [inline test/e2e/scheduling/events.go:22]: This belongs into the import section below.

Are you sure that it is needed? The file already had a `framework.ExpectEqual` call.

Same in other files.

File: test/e2e/scheduling/events.go
Comment (pohly) [inline test/e2e/scheduling/events.go:87]: `Failf` takes a format string, not a list of strings.

Also, `e` will be nil if `ok` is false. The `gingko.By` should crash. We need to move the check before it and *not* print `e`. Instead, use `framework.Failf("Expected *v1.Event, got %T %v", obj, obj)`.

Probably this never happens in practice, but we should still fix the test.

File: test/e2e/scheduling/predicates.go
Comment (pohly) [inline test/e2e/scheduling/predicates.go:131]: `Failf` with format string.

The message is misleading. The node was found, it just didn't have information about allocatable storage. Printing `node.Status` would be useful in this case.

File: test/e2e/scheduling/predicates.go
Comment (pohly) [inline test/e2e/scheduling/predicates.go:274]: Use `framework.Fail` when the parameter is a fixed string.

File: test/e2e/scheduling/predicates.go
Comment (pohly) [inline test/e2e/scheduling/predicates.go:348]: Format string.

File: test/e2e/scheduling/predicates.go
Comment (pohly) [inline test/e2e/scheduling/predicates.go:947]: Format string.

File: test/e2e/scheduling/preemption.go
Comment (pohly) [inline test/e2e/scheduling/preemption.go:98]: You can simplify the expression.

```suggestion
			if err != nil && !apierrors.IsAlreadyExists(err) {
```

File: test/e2e/scheduling/preemption.go
Comment (pohly) [inline test/e2e/scheduling/preemption.go:207]: Wrong failure message?

File: test/e2e/scheduling/preemption.go
Comment (pohly) [inline test/e2e/scheduling/preemption.go:312]: Same here.

File: test/e2e/scheduling/preemption.go
Comment (pohly) [inline test/e2e/scheduling/preemption.go:531]: You can simplify the expression.

File: test/e2e/scheduling/preemption.go
Comment (pohly) [inline test/e2e/scheduling/preemption.go:697]: And here.

File: test/e2e/scheduling/priorities.go
Comment (pohly) [inline test/e2e/scheduling/priorities.go:417]: Format string.

File: test/e2e/scheduling/priorities.go
Comment (pohly) [inline test/e2e/scheduling/priorities.go:423]: Format string.

File: test/e2e/scheduling/priorities.go
Comment (pohly) [inline test/e2e/scheduling/priorities.go:516]: Same as above, the failure message needs to be updated.

File: test/e2e/scheduling/priorities.go
Comment (pohly) [inline test/e2e/scheduling/priorities.go:527]: Format string.

**Review by pohly — CHANGES_REQUESTED** [review]
Can you check that the code compiles and passes govet before we trigger CI checks? You can do that with "make vet".

If your editor doesn't do it automatically, `hack/update-gofmt.sh` is useful for fixing formatting.

Also, I personally prefer to get fixes added to the existing commit (`git commit --amend`) over squashing at the end of the review. The contributor docs ask for separate commits, but IMHO that's more work and not useful anymore.

File: test/e2e/scheduling/events.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/events.go:22]: I guess, it was automatically added by Visual Studio Code, but I checked it and found that, it isn't needed. So removed this line of code from all the files.

File: test/e2e/scheduling/events.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/events.go:87]: Noted!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:131]: How about this?
`framework.Failf("unable to allocate to storage, node.Status: %v", node.Status)`

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:274]: Done!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:947]: Done!

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go:98]: Done!

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go:207]: I had deducted this result due to code above it, at line 198 i.e : 
`podPreempted := (err != nil && apierrors.IsNotFound(err)) ||
			  (err == nil && preemptedPod.DeletionTimestamp != nil)`

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go:531]: Done!

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go:697]: Done!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:274]: I found similar issue due to warning in Visual Studio Code at [https://github.com/Ahmed-Aghadi/kubernetes/blob/codeEnhanceNode/test/e2e/scheduling/predicates.go#L998](url)
i.e : 
`ginkgo.By(fmt.Sprintf("Running RC which reserves host port"))`
So I can change it to :
`ginkgo.By(fmt.Sprint("Running RC which reserves host port"))`
and same for : 
 [https://github.com/Ahmed-Aghadi/kubernetes/blob/codeEnhanceNode/test/e2e/scheduling/predicates.go#L1016](url)

Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:274]: And Should I change code at : 
 [https://github.com/Ahmed-Aghadi/kubernetes/blob/codeEnhanceNode/test/e2e/scheduling/predicates.go#L151](url) 
i.e : 
`var ephemeralStoragePerPod int64`

`ephemeralStoragePerPod = nodeMaxAllocatable / maxNumberOfPods`

to : 

`var ephemeralStoragePerPod int64 = nodeMaxAllocatable / maxNumberOfPods`

So we can merge variable declaration with assignment on the same line.

Comment (pohly) [reply test/e2e/scheduling/predicates.go]: This is shortest:
```
ephemeralStoragePerPod := nodeMaxAllocatable / maxNumberOfPods
```

Comment (pohly) [reply test/e2e/scheduling/predicates.go]: And the other line should be
```
ginkgo.By("Running RC which reserves host port")
```

File: test/e2e/scheduling/predicates.go
Comment (pohly) [reply test/e2e/scheduling/predicates.go]: This isn't about allocating storage.

The `node.Status.Allocatable[v1.ResourceEphemeralStorage]` map lookup checks whether such an entry exists in the map. I would just say `"node.Status.Allocatable %v does not contain entry %v", node.Status.Allocatable, v1.ResourceEphemeralStorage`

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go:206]: Am I correct?

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

**Comment by Ahmed-Aghadi** [comment]
@pohly I had updated necessary changes. Let me know if any further changes are required.

File: test/e2e/scheduling/preemption.go
Comment (pohly) [reply test/e2e/scheduling/preemption.go:206]: `podPreempted` is more than that "not found", there's also the `preemptedPod.DeletionTimestamp != nil` part.

I would say `"expected pod to be preempted, instead got pod %+v and error %v", preemptedPod, err`.

File: test/e2e/scheduling/preemption.go
Comment (pohly) [reply test/e2e/scheduling/preemption.go:311]: See above.

**Review by pohly — CHANGES_REQUESTED** [review]
Almost there :smile:  Two places need to be updated.

Can you squash all of your commits into one with a single commit message according to https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines?

**Comment by Ahmed-Aghadi** [comment]
@pohly I had squashed all the commits in one commit and run "make vet" and didn't got any error (atthough first I was getting error as my kubernetes folder wasn't inside folder named k8s.io) 
after running "make vet", a process named gopls started and its eating a lot of memory even after make vet had ran and finished without any error and without any output in terminal. What should I do next?

**Comment by pohly** [comment]
>  I had squashed all the commits in one commit and run "make vet" and didn't got any error (atthough first I was getting error as my kubernetes folder wasn't inside folder named k8s.io)

Good.

/ok-to-test

> after running "make vet", a process named gopls started and its eating a lot of memory even after make vet had ran and finished without any error and without any output in terminal.

I doubt that "make vet" started gopls. That is the Go language server, something that is probably used by your Visual Studio as backend for Go code.

> What should I do next?

You can pick some other directory from https://github.com/kubernetes/kubernetes/issues/105678 to work on.

**Comment by Ahmed-Aghadi** [comment]
> I doubt that "make vet" started gopls. That is the Go language server, something that is probably used by your Visual Studio as backend for Go code.

Should I use any other IDE for golang then?

**Comment by pohly** [comment]
I don't have a recommendation for another IDE on Windows.

**Comment by Ahmed-Aghadi** [comment]
> I don't have a recommendation for another IDE on Windows.

I am using ubuntu.

**Comment by Ahmed-Aghadi** [comment]
@pohly all the checks have successfully completed except one named "tide". So am I supposed to do anything regarding this?

**Comment by pohly** [comment]
/lgtm

/assign @ahg-g @ehashman

For approval.

**Comment by pohly** [comment]
> I am using ubuntu.

With Visual Studio? I am using Emacs with lsp mode. That also uses gopls, so it wouldn't be different.

> all the checks have successfully completed except one named "tide". So am I supposed to do anything regarding this?

That's normal. "Tide" is bot that merges PRs when they are ready and this one isn't quite there yet. I reviewed it, but code owners still need to approve it.

In addition, Kubernetes is now in code freeze and this change shouldn't be forced into 1.23. We'll just have to wait until the master branch opens for general enhancements like this.

**Comment by Ahmed-Aghadi** [comment]
> With Visual Studio? I am using Emacs with lsp mode. That also uses gopls, so it wouldn't be different.

I found that there is issue with gopls, where it uses a lot of memory [https://github.com/golang/go/issues/37670](url)
Although by opening just a specific folder and not whole kubernetes folder, It doesn't consumes a lot of memory ( most probably )

**Comment by SergeyKanzhelev** [comment]
/release-note-none
/kind cleanup
/priority backlog

not doing /triage accepted as this only a single Node-related test.

File: test/e2e/node/runtimeclass.go
Comment (SergeyKanzhelev) [inline test/e2e/node/runtimeclass.go:61]: will `apierrors.IsForbidden(err)` return `false` for `nil`?

Comment (pohly) [reply test/e2e/node/runtimeclass.go:61]: Yes. It uses errors.As under the hood. None of the API comments talk about nil directly, but the semantic is covered (nil is not assignable to a non-nil status type, so errors.As returns false) and the implementation checks for it.

Comment (pohly) [reply test/e2e/node/runtimeclass.go:61]: Other code is playing it safe and explicitly checks for nil beforehand, but that's unnecessary.

Comment (pohly) [reply test/e2e/node/runtimeclass.go:61]: I'm tempted to submit a patch both to k/k and the standard "errors" package with an explicit comment clarification...

Comment (pohly) [reply test/e2e/node/runtimeclass.go:61]: https://github.com/kubernetes/kubernetes/pull/106646

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

for runtimeclass test

**Comment by Ahmed-Aghadi** [comment]
@pohly what should I do next to make this pull request merged?

**Comment by pohly** [comment]
/assign @ahg-g

For approval.

**Comment by dims** [comment]
/assign @alculquicondor @Huang-Wei

NOTE: for sig-scheduling approval, this already has reviews and lgtm, so should be quick

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:944]: It would be helpful to have a different message for Failure vs Success.

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:1084]: log a Pod namespace/name instead of the whole object.

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:1086]: `ExpectEqual` accepts an explanation message. Add one

**Review by alculquicondor — COMMENTED** [review]
Thanks for these improvements

**Comment by pohly** [comment]
@Ahmed-Aghadi : can you address the review comments?

File: test/e2e/scheduling/priorities.go
Comment (Huang-Wei) [inline test/e2e/scheduling/priorities.go:417]: Including the node name would be helpful for debugging: (applies elsewhere)

```suggestion
			framework.Failf("Node %v: node.Status.Allocatable %v does not contain entry %v", node.Name, node.Status.Allocatable, v1.ResourceMemory)
```

File: test/e2e/scheduling/preemption.go
Comment (Huang-Wei) [inline test/e2e/scheduling/preemption.go:205]: for readability, prefer to move it under L198.

File: test/e2e/scheduling/predicates.go
Comment (Huang-Wei) [reply test/e2e/scheduling/predicates.go:1086]: Convert it to the `if <condition> then framework.Failf()` style?

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:946]: Done!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

File: test/e2e/scheduling/preemption.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/preemption.go]: Done!

File: test/e2e/scheduling/priorities.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/priorities.go]: Done!

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [reply test/e2e/scheduling/predicates.go:946]: "predicate" is an implementation detail of this function. That's not what we want to log.

We want to know whether there was a Success event (when `expectSuccess=true`) or a Failure event (when `expectSuccess=false`).

Comment (Huang-Wei) [reply test/e2e/scheduling/predicates.go:946]: @Ahmed-Aghadi I think this is still outstanding. Can you address this?

Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:946]: @alculquicondor  so should I replace this with : 
`framework.Logf("event was successful after performing the supplied action for pod %v", podName)`
and 
`framework.Logf("event  was failed after performing the supplied action for pod %v", podName)`

Comment (alculquicondor) [reply test/e2e/scheduling/predicates.go:946]: There shouldn't be `Logf`, just `Failf` if we were expecting Success but got Failed, or we were expecting Failed but got Success.

Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: @alculquicondor Done!

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:943]: This still doesn't seem correct.

I think what is causing confusion is line 941. The variable should be `observed` or `found` instead of `success`.

The message should be based on `expectSuccess`.

Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: @alculquicondor  what about now?

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:946]:
```suggestion
	if !expectSuccess && !success {
```

That's why I say that the name of the variable `success` is confusing. Can you rename it to `observed`?

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:944]:
```suggestion
		framework.Failf("Did not observe success event after performing the supplied action for pod %v", podName)
```

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go:946]: Sure!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

File: test/e2e/scheduling/predicates.go
Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

File: test/e2e/scheduling/predicates.go
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go:947]:
```suggestion
		framework.Failf("Did not observe failed event after performing the supplied action for pod %v", podName)
```

Comment (Ahmed-Aghadi) [reply test/e2e/scheduling/predicates.go]: Done!

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**
